internal/pkg/agent/application/upgrade/marker_watcher.go

```diff
@@ -0,0 +1,124 @@
+
+	"github.com/fsnotify/fsnotify"
+)
+
+type MarkerWatcher interface {
+	Watch() <-chan UpdateMarker
+	Errors() <-chan error
+	Run(ctx context.Context) error
+	Close() error
+}
+
+type MarkerFileWatcher struct {
+	watcher        *fsnotify.Watcher
+	markerFilePath string
+
+	logger *logger.Logger
+
+	updateCh chan UpdateMarker
+	errCh    chan error
+}
+
+func newMarkerFileWatcher(upgradeMarkerFilePath string, logger *logger.Logger) (MarkerWatcher, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create upgrade marker watcher: %w", err)
+	} // Watch the upgrade marker file's directory, not the file itself, so we
+
+	// notice the file even if it's deleted and recreated.
+	upgradeMarkerDirPath := filepath.Dir(upgradeMarkerFilePath)
+	err = watcher.Add(upgradeMarkerDirPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set watch on upgrade marker's directory [%s]: %w", upgradeMarkerDirPath, err)
+	}
+
+	return &MarkerFileWatcher{
+		watcher:        watcher,
+		markerFilePath: upgradeMarkerFilePath,
+		logger:         logger,
+		updateCh:       make(chan UpdateMarker),
+		errCh:          make(chan error),
+	}, nil
+}
+
+func (mfw *MarkerFileWatcher) Watch() <-chan UpdateMarker {
+	return mfw.updateCh
+}
+
+func (mfw *MarkerFileWatcher) Errors() <-chan error {
+	return mfw.errCh
+}
+
+func (mfw *MarkerFileWatcher) Run(ctx context.Context) error {
+	// Handle watching
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case err, ok := <-mfw.watcher.Errors:
+				mfw.logger.Debug("after there are watch errors")
+				if !ok { // Channel was closed (i.e. Watcher.Close() was called).
+					mfw.logger.Debug("upgrade marker watch's error channel was closed")
+					return
+				}
+				mfw.errCh <- fmt.Errorf("upgrade marker watch returned error: %w", err)
+				continue
+			case e, ok := <-mfw.watcher.Events:
+				mfw.logger.Debugf("after there is a watch event: [%s]", e.String())
+				if !ok { // Channel was closed (i.e. Watcher.Close() was called).
+					mfw.logger.Debug("upgrade marker watch's events channel was closed")
+					return
+				}
+
+				if e.Name != mfw.markerFilePath {
+					// Event is for a file other than the upgrade marker; ignore it.
+					continue
+				}
+
+				switch e.Op {
+				case fsnotify.Create, fsnotify.Write:
+					// Upgrade marker file was created or updated; read its contents
+					// and send them over the update channel.
+
+					marker, err := loadMarker(mfw.markerFilePath)
+					if err != nil {
+						mfw.errCh <- fmt.Errorf("unable to load upgrade marker from watch: %w", err)
+						return
+					}
+
+					mfw.updateCh <- *marker
+				}
+			}
+		}
+	}()
+
+	// Do an initial read from the upgrade marker file, in case the file
+	// is already present before the watching starts.
+	marker, err := loadMarker(mfw.markerFilePath)
+	if err != nil {
+		return fmt.Errorf("unable to load upgrade marker from watch: %w", err)
```

Comment:
+1

Reply:
When I wrote this code here I did inspect `loadMarker` to ensure that [it returned `nil, nil` if the file was not present](https://github.com/elastic/elastic-agent/blob/72653ac2b0b57deedd276c3abc8d19f9085b3154/internal/pkg/agent/application/upgrade/step_mark.go#L156-L158).  So if the file does not exist yet or is still being written to, `loadMarker` here will not return an error (which is what we want).

So the only errors that could be returned from `loadMarker` are:
- if there's some error reading the file into memory, or
- if there's some error parsing the contents of the file as YAML into memory

I think the question is whether either of these errors could be transient.  That is, if a call to `loadMarker` were to be attempted again (as part of the goroutine code), could these errors go away or will they always be returned. 

I can't come up with conditions that would make the first error transient but perhaps there are some.

As for the second error, I suppose there's some scenario in which the file is still being written to so only part of the contents have been written. So that would make the second error transient.

To address the (potentially) transient nature of these errors, I'm good with removing the error check here and just checking if the returned `marker` is `nil` (which we are doing already in the code).

Reply:
I will address this as part of the refactoring for https://github.com/elastic/elastic-agent/pull/3694#discussion_r1386317575.

Reply:
> I will address this as part of the refactoring for [#3694 (comment)](https://github.com/elastic/elastic-agent/pull/3694#discussion_r1386317575).

Addressed as part of the refactoring done in 83577da728.